ENH: Use a faster t-SNE implementation

Currently, scikit-learn's implementation is used. One of the following would likely perform better in terms of runtime:
- https://github.com/DmitryUlyanov/Multicore-TSNE
- https://github.com/pavlin-policar/openTSNE
- https://github.com/lvdmaaten/bhtsne

The current lines that would need to be changed are: 
https://github.com/zoj613/pyloras/blob/b896367dac4e970bb31d2b4cac69fe26f7627a2d/pyloras/_loras.py#L112-L115
https://github.com/zoj613/pyloras/blob/b896367dac4e970bb31d2b4cac69fe26f7627a2d/pyloras/_loras.py#L149

Hi,
I found that out of the three  MultiCore-TSNE from  [https://github.com/DmitryUlyanov/Multicore-TSNE](https://github.com/DmitryUlyanov/Multicore-TSNE) is the fastest and relatively simple to use. It can directly replace `from sklearn.manifold import TSNE`. But there is an install error being faced while using pip to install MultiCore-TSNE.

```
Requirement already satisfied: numpy in c:\users\username\appdata\local\programs\python\python39\lib\site-packages (from MulticoreTSNE==0.1) (1.21.0)
Requirement already satisfied: cffi in c:\users\username\appdata\local\programs\python\python39\lib\site-packages (from MulticoreTSNE==0.1) (1.14.5)
Requirement already satisfied: pycparser in c:\users\username\appdata\local\programs\python\python39\lib\site-packages (from cffi->MulticoreTSNE==0.1) (2.20)
Building wheels for collected packages: MulticoreTSNE
  Building wheel for MulticoreTSNE (setup.py) ... error
  ERROR: Command errored out with exit status 1:
   command: 'c:\users\username\appdata\local\programs\python\python39\python.exe' -u -c 'import io, os, sys, setuptools, tokenize; sys.argv[0] = '"'"'C:\\Users\\username\\AppData\\Local\\Temp\\pip-req-build-kpt7ghlg\\setup.py'"'"'; __file__='"'"'C:\\Users\\username\\AppData\\Local\\Temp\\pip-req-build-kpt7ghlg\\setup.py'"'"';f = getattr(tokenize, '"'"'open'"'"', open)(__file__) if os.path.exists(__file__) else io.StringIO('"'"'from setuptools import setup; setup()'"'"');code = f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' bdist_wheel -d 'C:\Users\username\AppData\Local\Temp\pip-wheel-69r1ew1b'
       cwd: C:\Users\username\AppData\Local\Temp\pip-req-build-kpt7ghlg\
  Complete output (28 lines):
  running bdist_wheel
  running build
  running build_py
  creating build
  creating build\lib.win-amd64-3.9
  creating build\lib.win-amd64-3.9\MulticoreTSNE
  copying MulticoreTSNE\__init__.py -> build\lib.win-amd64-3.9\MulticoreTSNE
  creating build\lib.win-amd64-3.9\MulticoreTSNE\tests
  copying MulticoreTSNE\tests\test_base.py -> build\lib.win-amd64-3.9\MulticoreTSNE\tests
  copying MulticoreTSNE\tests\__init__.py -> build\lib.win-amd64-3.9\MulticoreTSNE\tests
  running egg_info
  creating MulticoreTSNE.egg-info
  writing MulticoreTSNE.egg-info\PKG-INFO
  writing dependency_links to MulticoreTSNE.egg-info\dependency_links.txt
  writing requirements to MulticoreTSNE.egg-info\requires.txt
  writing top-level names to MulticoreTSNE.egg-info\top_level.txt
  writing manifest file 'MulticoreTSNE.egg-info\SOURCES.txt'
  reading manifest file 'MulticoreTSNE.egg-info\SOURCES.txt'
  reading manifest template 'MANIFEST.in'
  writing manifest file 'MulticoreTSNE.egg-info\SOURCES.txt'
  running build_ext
  cmake version 3.20.5

  CMake suite maintained and supported by Kitware (kitware.com/cmake).
  CMake Error: Unknown argument --
  CMake Error: Run 'cmake --help' for all supported options.

  ERROR: Cannot generate Makefile. See above errors.
  ----------------------------------------

```
In case, this error is negligible during actual implementation. Then I have changed the code accordingly to fit MultiCore-TSNE library.

--- a/README.md
+++ b/README.md
@@ -45,12 +45,17 @@ X, y = make_classification(n_samples=20000, n_features=5, n_informative=5,
                            weights=[0.01, 0.05, 0.94],
                            class_sep=0.8, random_state=0)
 
-lrs = LORAS(random_state=0, embedding_params={'perplexity': 35, 'n_iter': 250})
+lrs = LORAS(random_state=0, manifold_learner_params={'perplexity': 35, 'n_iter': 250})
 print(sorted(Counter(y).items()))
 # [(0, 270), (1, 1056), (2, 18674)]
 X_resampled, y_resampled = lrs.fit_resample(X, y)
 print(sorted(Counter(y_resampled.astype(int)).items()))
 # [(0, 18674), (1, 18674), (2, 18674)]
+
+# one can also use any custom 2d manifold learner via the ``manifold_learner` parameter
+from umap import UMAP
+LORAS(manifold_learner=UMAP()).fit_resample(X, y)
+
 ```
 
 ## Visualization

--- a/pyloras/_loras.py
+++ b/pyloras/_loras.py
@@ -53,11 +53,18 @@ class LORAS(BaseOverSampler):
         between ``2`` and the number of features used in the fitting data.
         If not given, the value will be set to the total number of features in
         fitting data.
-    embedding_params : dict, default=None
-        A dictionary of additional parameters to pass to the
-        :class:`~sklearn.manifold.TSNE` object when creating a 2D manifold of
-        the fitting data. The keys are the parameter names and the keys are the
-        values. If not given, the default values are used.
+    manifold_learner : object, default=None
+        An instance of an object that to perform a 2-dimensional embedding of
+        a dataset. It must implement the scikit-learn Estimator interface,
+        ``fit_transform`` and ``set_params`` methods must be implemented.
+        If not given, the :class:`~sklearn.manifold.TSNE` class is used to
+        obtain the 2d manifold of the data. Defaults to None.
+    manifold_learner_params : dict, default=None
+        A dictionary of additional parameters to pass to the instance of the
+        ``manifold_learner`` (or TSNE if ``manifold_learner`` is None) when
+        creating a 2D manifold of the fitting data. The keys are the parameter
+        names and the values are the values. If not given, the default values
+        are used.
     {random_state}
     {n_jobs}
 
@@ -91,7 +98,8 @@ class LORAS(BaseOverSampler):
         n_shadow=None,
         std=0.005,
         n_affine=None,
-        embedding_params=None,
+        manifold_learner=None,
+        manifold_learner_params=None,
         random_state=None,
         n_jobs=None
     ):
@@ -100,18 +108,35 @@ class LORAS(BaseOverSampler):
         self.n_shadow = n_shadow
         self.std = std
         self.n_affine = n_affine
-        self.embedding_params = embedding_params
+        self.manifold_learner = manifold_learner
+        self.manifold_learner_params = manifold_learner_params
         self.random_state = random_state
         self.n_jobs = n_jobs
+
+    def _check_2d_manifold_learner(self):
+        if (not hasattr(self.manifold_learner, "fit_transform") or
+                not hasattr(self.manifold_learner, "set_params")):
+            raise ValueError(
+                "The 2d manifold learner must implement the ``fit_transform`` "
+                "and ``set_params`` methods"
+            )
 
     def _initialize_params(self, X, y, rng):
         """Initialize the parameter values to their appropriate values."""
         f_size = X.shape[1]
         self.n_affine_ = f_size if self.n_affine is None else self.n_affine
-        self.tsne_ = TSNE(n_components=2, random_state=rng)
 
-        if self.embedding_params is not None:
-            self.tsne_.set_params(**self.embedding_params)
+        if self.manifold_learner:
+            self._check_2d_manifold_learner()
+            self.manifold_learner_ = self.manifold_learner
+        else:
+            self.manifold_learner_ = TSNE(n_components=2)
+        if self.manifold_learner_params is not None:
+            self.manifold_learner_.set_params(**self.manifold_learner_params)
+        try:
+            self.manifold_learner_.set_params(random_state=rng)
+        except ValueError:
+            pass
 
         _, y_counts = np.unique(y, return_counts=True)
         if self.n_neighbors is None:
@@ -153,7 +178,7 @@ class LORAS(BaseOverSampler):
             if samples_to_make == 0:
                 continue
             X_minority = X[y == minority_class]
-            X_embedded = self.tsne_.fit_transform(X_minority)
+            X_embedded = self.manifold_learner_.fit_transform(X_minority)
             self.nn_.fit(X_embedded)
             neighborhoods = self.nn_.kneighbors(X_embedded, return_distance=False)
             num_loras = ceil(samples_to_make / X_embedded.shape[0])

--- a/tests/test_loras.py
+++ b/tests/test_loras.py
@@ -1,6 +1,7 @@
 import pytest
 import numpy as np
 from sklearn.neighbors import NearestNeighbors
+from sklearn.base import BaseEstimator
 
 from pyloras import LORAS
 
@@ -76,9 +77,29 @@ def test_loras(data):
     with pytest.raises(ValueError):
         LORAS(n_affine=500).fit_resample(X, y)
 
-    lrs = LORAS(random_state=rng, embedding_params={'n_iter': 270})
+    lrs = LORAS(random_state=rng, manifold_learner_params={'n_iter': 270})
     X_res, y_res = lrs.fit_resample(X, y)
-    assert lrs.tsne_.n_iter == 270
+    assert lrs.manifold_learner_.n_iter == 270
+
+    # test for using a custom 2d manifold learner other than sklearn's TSNE.
+    class WrongCustomEmbedding(BaseEstimator):
+        def __init__(self, n_components=None):
+            self.n_components = None
+
+    class RightCustomEmbedding(WrongCustomEmbedding):
+        def fit_transform(self, X, y=None):
+            self.n_components_ = self.n_components if self.n_components else 2
+            return X[:, :2]
+
+    lrs = LORAS(manifold_learner=WrongCustomEmbedding())
+    with pytest.raises(ValueError):
+        lrs.fit_resample(X, y)
+
+    lrs = LORAS(
+        manifold_learner=RightCustomEmbedding(),
+        manifold_learner_params={'n_components': 2}
+    )
+    X_res, y_res = lrs.fit_resample(X, y)
 
     # test if no samples are added for an already balanced dataset.
     y = np.array([1., 1, 1., 1., 1., 1., 1., 1., 1, 1., 1., 1., 1., 1., 1.,
@@ -86,4 +107,3 @@ def test_loras(data):
     X_res, y_res = lrs.fit_resample(X, y)
     np.testing.assert_allclose(X_res, X)
     np.testing.assert_allclose(y_res, y)
-

